src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs

```diff
@@ -205,9 +209,39 @@ private ImmutableArray<ISymbol> GetSymbolsForTypeOrNamespaceContext()
                 return symbols.WhereAsArray(s => !s.IsDelegateType() && !s.IsInterfaceType());
             }
 
+            if (_context.IsAsyncMemberDeclarationContext)
+            {
+                return symbols.WhereAsArray(IsSymbolValidForAsyncDeclarationContext);
+            }
+
             return symbols;
         }
 
+        private bool IsSymbolValidForAsyncDeclarationContext(ISymbol symbol)
+        {
+            if (symbol.IsNamespace())
+            {
+                return true;
+            }
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType())
+            {
+                return false;
+            }
+
+            if (namedType.TypeKind == TypeKind.Interface)
+            {
+                // The only interfaces, that are valid in async context are IAsyncEnumerable and IAsyncEnumerator.
+                // So if we are validating an interface, then we can just check for 2 of this possible variants
+                return namedType == _context.SemanticModel.Compilation.IAsyncEnumerableOfTType() ||
+                       namedType == _context.SemanticModel.Compilation.IAsyncEnumeratorOfTType();
```

Comment:
named types need to use .Equals for equality comparisons, they're not guaranteed to have reference equality.